src/invariant/InvariantManager.h

```diff
@@ -0,0 +1,51 @@
+#pragma once
+
+// Copyright 2017 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "herder/TxSetFrame.h"
+#include <memory>
+
+namespace stellar
+{
+
+class Application;
+class Invariant;
+class LedgerDelta;
+
+/**
+ * InvariantManager maintains a registry of available invariants and
+ * supports enabling them dynamically, such as at configuration time.
+ * When the appropriate event, such as a ledger close, triggers the
+ * InvariantManager it will check each of the enabled invariants and
+ * throw InvariantDoesNotHold if any are violated.
+ */ 
+class InvariantManager
+{
+  public:
+    static std::unique_ptr<InvariantManager> create(Application& app);
+
+    virtual ~InvariantManager()
+    {
+    }
+
+    virtual void
+    checkOnLedgerClose(TxSetFramePtr const& txSet, LedgerDelta const& delta)
```

Comment:
I think that should be abstract as well (= 0)

Reply:
I think that if we use the `override` keyword in the actual invariants implementation we'll get both of both worlds: ability to add new invariant types without having to update the unrelated invariants. I think that is what Jon is going for.

Reply:
I agree with @vogel. In Invariant we should have the checkOn* functions be virtual with default implementation as @MonsieurNicolas mentions. But in the InvariantManager I think we should have the checkOn* be pure virtual, since any reasonable implementation of an InvariantManager should override these methods to at least call into the Invariants that it manages. This is definitely safer than letting people accidentally choose the default (meaning do nothing) implementation.